pkg/endpoint/endpoint.go

```diff
@@ -1962,6 +1970,20 @@ func (e *Endpoint) SetStateLocked(toState, reason string) bool {
 OKState:
 	e.state = toState
 	e.logStatusLocked(Other, OK, reason)
+
+	// Initial state transitions i.e nil --> waiting-for-identity
+	// need to be handled correctly while updating metrics.
+	if fromState != "" {
```

Comment:
This only works for new endpoints. If we restore endpoints, the fromState will be "" but there is no previous increment operation to compensate for. The code as-is will result in negative state counters for restore endpoints when Cilium is restarted. I think you have to special case the restore operation as the previous state as recovered in `ParseEndpoint()` is not represented in the in-memory metrics. This also means that you can undo the additions to the valid state transitions above.

Reply:
You mean I need to add a special case for 

Would this explicit check make more sense then?

```
       // Initial state transitions i.e nil --> waiting-for-identity
	// need to be handled correctly while updating metrics.
	// Note that if we are transitioning from some state to restoring
	// state, we cannot decrement the old state counters as they will not
	// be accounted for in the metrics.
	if fromState != "" && toState != StateRestoring {
		metrics.EndpointStateCount.
			WithLabelValues(fromState).Dec()
	}
```

Reply:
I am unclear about the second half of the comment, but the reason I have added StateRestoring in the state transitions above in the switch case is to make the code jump to `OKState:` because we still need to execute this part of the code for ` StateRestoring`
```
// Since StateDisconnected is the final state, after which the
	// endpoint is gone, we should not increment metrics for this state.
	if toState != "" && toState != StateDisconnected {
		metrics.EndpointStateCount.
			WithLabelValues(toState).Inc()
	}
```
 i.e increment the `StateRestoring` part? Let me know if my understanding is correct here.. I hope I have interpreted your review comment correctly.

Reply:
done.. not sure if it addresses all your concerns, please let me know if I have not understood something.